File: tests/benchmark/pooling/bench_adaptive_avg_pool2d.py
"""
AdaptiveAvgPool2d算子性能测试
测试PyTorch的adaptive_avg_pool2d操作性能
"""

import sys
from pathlib import Path

# 添加common目录到路径
common_dir = Path(__file__).parent.parent / 'common'
if str(common_dir) not in sys.path:
    sys.path.insert(0, str(common_dir))

from typing import List, Tuple
import torch
import torch.nn.functional as F
from benchmark_framework import BenchmarkFramework, BenchmarkConfig
from timer import Timer


class AdaptiveAvgPool2dBenchmark(BenchmarkFramework):
    """AdaptiveAvgPool2d算子基准测试类"""
    
    def get_required_shapes_count(self) -> int:
        """AdaptiveAvgPool2d算子只需要一个shape（4D输入）"""
        return 1
    
    def run_benchmark(self, config: BenchmarkConfig) -> float:
        """执行PyTorch的adaptive_avg_pool2d操作性能测试"""
        if len(config.shapes) != 1:
            raise ValueError(f"AdaptiveAvgPool2d benchmark requires exactly 1 shape, got {len(config.shapes)}")
        
        shape = config.shapes[0]
        device = self._parse_device(config.device)
        dtype = torch.float32 if config.dtype == 'float32' else torch.float64
        
        # AdaptiveAvgPool2d需要4D输入 (N, C, H, W)
        if len(shape) != 4:
            raise ValueError(f"AdaptiveAvgPool2d requires 4D shape (N, C, H, W), got {shape}")
        
        x = torch.ones(shape, dtype=dtype, device=device)
        
        # 使用默认的输出大小 (1, 1)
        output_size = (1, 1)
        
        # 预热
        for _ in range(config.warmup_cnt):
            result = F.adaptive_avg_pool2d(x, output_size)
        if device.type == 'cuda':
            torch.cuda.synchronize()
        
        # 正式测试
        timer = Timer()
        timer.start()
        
        for _ in range(config.repeat_cnt):
            result = F.adaptive_avg_pool2d(x, output_size)
        if device.type == 'cuda':
            torch.cuda.synchronize()
        
        return timer.elapsed_us() / config.repeat_cnt
    
    def get_default_shapes(self):
        """获取默认测试形状"""
        return [
            [(1, 1, 3, 3)],
            [(1, 3, 10, 10)],
            [(1, 3, 32, 32)],
            [(1, 64, 64, 64)],
            [(4, 3, 224, 224)],
        ]
    
    def validate_shapes(self, shapes: List[Tuple[int, ...]]) -> None:
        """验证形状是否有效"""
        if len(shapes) != 1:
            raise ValueError(f"AdaptiveAvgPool2d requires exactly 1 shape, got {len(shapes)}")
        if len(shapes[0]) != 4:
            raise ValueError(f"AdaptiveAvgPool2d requires 4D shape (N, C, H, W), got {shapes[0]}")


def benchmark_adaptive_avg_pool2d_comparison(device_filter=None,
                             shape_filter=None,
                             warmup_cnt=5,
                             repeat_cnt=100,
                             inplace=False,
                             verbose=False):
    """
    运行adaptive_avg_pool2d算子的PyTorch性能测试
    
    Args:
        device_filter: 设备过滤，'cpu'或'cuda'
        shape_filter: shape过滤，例如 '1,3,224,224'
        warmup_cnt: 预热次数
        repeat_cnt: 重复次数
        inplace: 是否使用就地操作（adaptive_avg_pool2d不支持inplace）
        verbose: 是否输出详细信息
    
    Returns:
        PyTorch测试结果列表
    """
    benchmark = AdaptiveAvgPool2dBenchmark()
    return benchmark.run(
        device_filter=device_filter,
        shape_filter=shape_filter,
        warmup_cnt=warmup_cnt,
        repeat_cnt=repeat_cnt,
        inplace=False,  # adaptive_avg_pool2d不支持inplace
        verbose=verbose
    )


if __name__ == "__main__":
    benchmark = AdaptiveAvgPool2dBenchmark()
    sys.exit(benchmark.run_from_command_line())

File: tests/benchmark/pooling/bench_adaptive_avg_pool2d.cpp
#include <vector>
#include "benchmark/common/benchmark_framework.h"
#include "benchmark/common/timer.h"
#include "origin.h"
#include "origin/operators/pooling/adaptive_avg_pool2d.h"

using namespace origin;
using namespace origin::benchmark;
namespace F = origin::functional;

/**
 * @brief AdaptiveAvgPool2d算子基准测试类
 */
class AdaptiveAvgPool2dBenchmark : public BenchmarkFramework
{
public:
    double run_benchmark(const BenchmarkConfig &config) override
    {
        if (config.shapes.size() != 1)
        {
            THROW_RUNTIME_ERROR("AdaptiveAvgPool2d benchmark requires exactly 1 shape, got {}", config.shapes.size());
        }

        const Shape &shape = config.shapes[0];

        // AdaptiveAvgPool2d需要4D输入 (N, C, H, W)
        if (shape.ndims() != 4)
        {
            THROW_RUNTIME_ERROR("AdaptiveAvgPool2d requires 4D shape (N, C, H, W), got {}", shape.to_string());
        }

        size_t numel = shape.elements();
        std::vector<float> data(numel, 1.0f);

        auto x = Tensor(data, shape, origin::dtype(config.dtype).device(config.device));

        // 使用默认的输出大小 (1, 1)
        std::pair<int, int> output_size(1, 1);

        // 预热
        for (int i = 0; i < config.warmup_cnt; ++i)
        {
            auto result = F::adaptive_avg_pool2d(x, output_size);
        }
        if (config.device.type() == DeviceType::kCUDA)
        {
            cuda::synchronize();
        }

        // 正式测试
        Timer timer;
        timer.start();

        for (int i = 0; i < config.repeat_cnt; ++i)
        {
            auto result = F::adaptive_avg_pool2d(x, output_size);
        }
        if (config.device.type() == DeviceType::kCUDA)
        {
            cuda::synchronize();
        }

        return timer.elapsed_us() / config.repeat_cnt;
    }

    std::vector<std::vector<Shape>> get_default_shapes() const override
    {
        return {
            {Shape({1, 1, 3, 3})},    {Shape({1, 3, 10, 10})},   {Shape({1, 3, 32, 32})},
            {Shape({1, 64, 64, 64})}, {Shape({4, 3, 224, 224})},
        };
    }

    size_t get_required_shapes_count() const override { return 1; }

    void validate_shapes(const std::vector<Shape> &shapes) const override
    {
        if (shapes.size() != 1)
        {
            THROW_RUNTIME_ERROR("AdaptiveAvgPool2d requires exactly 1 shape, got {}", shapes.size());
        }
        if (shapes[0].ndims() != 4)
        {
            THROW_RUNTIME_ERROR("AdaptiveAvgPool2d requires 4D shape (N, C, H, W), got {}", shapes[0].to_string());
        }
    }

    std::string get_operator_name() const override { return "AdaptiveAvgPool2d"; }
};

int main(int argc, char *argv[])
{
    AdaptiveAvgPool2dBenchmark benchmark;
    return benchmark.run(argc, argv);
}
